In [1]:
import pandas as pd
import numpy as np
import darkchem
import math

Using TensorFlow backend.


#### Several reactions of Hydration: `Benzene` `1-Butene` `2-Butene` `1-Hexene`

In [2]:
# Load reactants 
smi_in = {'reactants':['c1ccccc1','CCC=C','C/C=C/C','CCCCC=C']}
# Load products  
smi_out = {'products':['Oc1ccccc1','CCC(O)C','CCC(O)C','CCCCC(O)C']}       
# Transform to DataFrame
smi_in = pd.DataFrame(smi_in)  
smi_out = pd.DataFrame(smi_out)
# Represent reactants in the form of vector
rct = [darkchem.utils.struct2vec(smi) for smi in smi_in['reactants']]  
# Represent prodcuts in the form of vector
prod = [darkchem.utils.struct2vec(smi) for smi in smi_out['products']] 

In [3]:
# Put those vectors into DataFrame
rct = pd.DataFrame(rct)
prod = pd.DataFrame(prod)

In [4]:
# Calcualte the lengh of each reaction path
s = 0
d = []
for i in range(4):
    for j in range(100):
        s += (rct.loc[i][j] - prod.loc[i][j])**2
    ds = np.sqrt(s)
    d.append(ds)
print('The lengh of each reaction path =', d)

The lengh of each reaction path = [18.05547008526779, 32.43454948045371, 34.20526275297414, 43.54308211415448]


In [5]:
# Calculate the dot product of each reaction
s2 = 0
d2= []
for i in range(4):
    for j in range(100):
        s2 += (rct.loc[i][j] * prod.loc[i][j])
    #ds = np.sqrt(s)
    d2.append(s2)


# Calculate the product of each reactant and product
s3 = 0
s4 = 0
d3= []
for i in range(4):
    for j in range(100):
        s3 += (rct.loc[i][j])**2
        s4 += (prod.loc[i][j])**2
    dd = np.sqrt(s3)
    dx = np.sqrt(s4)
    xx = dd*dx
    d3.append(xx)

# Insert the formula of Cosine theorem
a = []
for i in range(4):
    cos = d2[i]/d3[i]  #this is the value of cos(sita)
    angle = math.acos(cos)  #this is the value of sita
    a.append(angle)
print('The angle between the vectors of reactants and products of each reactions =')
print(a)

The angle between the vectors of reactants and products of each reactions =
[0.6115044811833035, 0.6457305942915628, 0.509857907016216, 0.5404779029054397]


#### Hydration
| Index |Benzene |1-Butene|2-Butene|1-Hexene|
|-------|--------|--------|--------|--------|
| Length|18.05547|32.43454|34.20526|43.54308|
| Angle |0.611504|0.645730|0.509857|0.540477|